cmd/runner.go

```diff
@@ -0,0 +1,142 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/spf13/cobra"
+
+	"github.com/giantswarm/retagger/pkg/config"
+	"github.com/giantswarm/retagger/pkg/registry"
+)
+
+type runner struct {
+	flag   *flag
+	logger micrologger.Logger
+	stdout io.Writer
+	stderr io.Writer
+}
+
+func (r *runner) Run(cmd *cobra.Command, args []string) error {
+	ctx := context.Background()
+
+	err := r.flag.Validate()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.run(ctx, cmd, args)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *runner) run(ctx context.Context, cmd *cobra.Command, args []string) error {
+	var err error
+
+	var conf *config.Config
+	{
+		conf, err = config.FromFile(r.flag.ConfigFile)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var destRegistry *registry.Registry
+	{
+		c := registry.Config{
+			Host:         r.flag.Host,
+			Organisation: r.flag.Organization,
+			Password:     r.flag.Password,
+			Username:     r.flag.Username,
+			LogFunc:      nil,
+		}
+		destRegistry, err = registry.New(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	err = destRegistry.Login()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, image := range conf.Images {
+		for _, tag := range image.Tags {
+			imageName := image.Name
+			if image.OverrideRepoName != "" {
+				r.logger.LogCtx(ctx, "message", "level", "debug", fmt.Sprintf("Override Name specified. Using %s as mirrored image name", image.OverrideRepoName))
+				imageName = image.OverrideRepoName
+			}
+			r.logger.LogCtx(ctx, "message", "level", "debug", fmt.Sprintf("managing: %v, %v, %v", imageName, tag.Sha, tag.Tag))
+
+			for _, customImage := range tag.CustomImages {
+				ok, err := destRegistry.CheckImageTagExists(imageName, tag.Tag)
+				if ok {
+					r.logger.LogCtx(ctx, "message", "level", "debug", fmt.Sprintf("rebuilt image %q with tag %q already exists, skipping", imageName, fmt.Sprintf("%s-%s", tag.Tag, customImage.TagSuffix)))
```

Comment:
Five levels of indentation is reaching the point of being difficult to read. I might break this inner loop into a separate function if it doesn't cause too much trouble.

Reply:
I want to move the code starting somewhere around `for _, image := range conf.Images {` into a different package with the next PR. Didn't do it now as this would bloat up this PR even more.

Reply:
Fine with me. Keep PRs small :)